File: clients/node/conformance/scenarios.ts
// Reading conformance/scenarios/*.json. The scenarios are the specification;
// this file only decodes them, and refuses a field it does not understand.

import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";

/**
 * A number as the scenario wrote it. The digits are kept so a whole number too
 * large for a double is still compared, and written back, exactly.
 */
export class Literal {
  constructor(readonly text: string) {}

  /** The value as protobuf-JSON: digits for a 64-bit integer, a number otherwise. */
  toJSON(): number | string {
    const value = Number(this.text);
    return Number.isInteger(value) && !Number.isSafeInteger(value) ? this.text : value;
  }
}

/**
 * The source a scenario needs parsed before its call: one fixture, or several
 * parsed together as one model, named by their fixtures.
 */
export interface ScenarioModel {
  fixture?: string;
  fixtures?: string[];
  language?: string;
  strict_conformance?: boolean;
}

/** What a call must answer. Every field is optional; an absent status means OK. */
export interface Expect {
  status?: string;
  status_message_contains?: string;
  response?: Record<string, unknown>;
  contains?: Record<string, string>;
  contains_all?: Record<string, string[]>;
  non_empty?: string[];
  absent?: string[];
  /** Values are Literal, because a scenario's numbers are kept as written. */
  counts?: Record<string, Literal>;
  min_counts?: Record<string, Literal>;
}

/** One conformance case: a call to make and what it must answer. */
export interface Scenario {
  id: string;
  description?: string;
  rpc: string;
  requires_capabilities?: string[];
  expect_without_capability?: Expect;
  model?: ScenarioModel;
  request?: Record<string, unknown>;
  expect?: Expect;
  /** The file it came from, for error messages. */
  file: string;
}

const SCENARIO_FIELDS = new Set([
  "id",
  "description",
  "rpc",
  "requires_capabilities",
  "expect_without_capability",
  "model",
  "request",
  "expect",
]);

const EXPECT_FIELDS = new Set([
  "status",
  "status_message_contains",
  "response",
  "contains",
  "contains_all",
  "non_empty",
  "absent",
  "counts",
  "min_counts",
]);

const MODEL_FIELDS = new Set(["fixture", "fixtures", "language", "strict_conformance"]);

/** The scenario's RPC as a bare method name. */
export function methodOf(scenario: Scenario): string {
  const slash = scenario.rpc.lastIndexOf("/");
  return slash === -1 ? scenario.rpc : scenario.rpc.slice(slash + 1);
}

/** Reads every scenario file in `dir`, in file then declaration order. */
export function loadScenarios(dir: string): Scenario[] {
  const files = readdirSync(dir)
    .filter((name) => name.endsWith(".json"))
    .sort();
  if (files.length === 0) {
    throw new Error(`no scenario files in ${dir}`);
  }
  const scenarios: Scenario[] = [];
  const seen = new Map<string, string>();
  for (const name of files) {
    const path = join(dir, name);
    const suite = parseKeepingLiterals(readFileSync(path, "utf8"));
    for (const scenario of scenariosOf(suite, path)) {
      const where = seen.get(scenario.id);
      if (where !== undefined) {
        throw new Error(`${path}: scenario id ${JSON.stringify(scenario.id)} is already declared in ${where}`);
      }
      seen.set(scenario.id, path);
      scenarios.push(scenario);
    }
  }
  return scenarios;
}

/** Parses a scenario file, keeping every number as the literal it was written as. */
function parseKeepingLiterals(text: string): unknown {
  return JSON.parse(text, function (_key: string, value: unknown, context?: { source?: string }) {
    const source = context?.source;
    return typeof value === "number" && source !== undefined ? new Literal(source) : value;
  }) as unknown;
}

function scenariosOf(suite: unknown, path: string): Scenario[] {
  if (!isObject(suite) || !Array.isArray(suite["scenarios"])) {
    throw new Error(`${path}: expected an object with a "scenarios" list`);
  }
  return suite["scenarios"].map((entry: unknown) => scenarioOf(entry, path));
}

function scenarioOf(entry: unknown, path: string): Scenario {
  if (!isObject(entry)) {
    throw new Error(`${path}: a scenario must be an object`);
  }
  reject(entry, SCENARIO_FIELDS, path, "a scenario");
  const id = entry["id"];
  const rpc = entry["rpc"];
  if (typeof id !== "string" || id === "" || typeof rpc !== "string" || rpc === "") {
    throw new Error(`${path}: every scenario needs an id and an rpc`);
  }
  const model = entry["model"];
  if (model !== undefined) {
    if (!isObject(model)) {
      throw new Error(`${path}: ${id}: model must be an object`);
    }
    reject(model, MODEL_FIELDS, path, `${id}: model`);
  }
  for (const key of ["expect", "expect_without_capability"] as const) {
    const expect = entry[key];
    if (expect !== undefined) {
      if (!isObject(expect)) {
        throw new Error(`${path}: ${id}: ${key} must be an object`);
      }
      reject(expect, EXPECT_FIELDS, path, `${id}: ${key}`);
    }
  }
  return { ...(entry as unknown as Omit<Scenario, "file">), file: path };
}

function reject(node: Record<string, unknown>, known: ReadonlySet<string>, path: string, what: string): void {
  for (const key of Object.keys(node)) {
    if (!known.has(key)) {
      throw new Error(`${path}: ${what} has no field ${JSON.stringify(key)}`);
    }
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: client/java/opensysml-conformance/src/main/java/org/openmbee/opensysml/conformance/Scenarios.java
package org.openmbee.opensysml.conformance;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;

/** Reads the scenario files, in file then declaration order. */
final class Scenarios {

  private static final String EXPECT = "expect";
  private static final String EXPECT_WITHOUT_CAPABILITY = "expect_without_capability";
  private static final String MODEL = "model";
  private static final String REQUEST = "request";
  private static final String RESPONSE = "response";
  private static final String STRICT_CONFORMANCE = "strict_conformance";

  private static final Set<String> SCENARIO_MEMBERS =
      Set.of(
          "id",
          "description",
          "rpc",
          "requires_capabilities",
          EXPECT_WITHOUT_CAPABILITY,
          MODEL,
          REQUEST,
          EXPECT);
  private static final Set<String> EXPECT_MEMBERS =
      Set.of(
          "status",
          "status_message_contains",
          RESPONSE,
          "contains",
          "contains_all",
          "non_empty",
          "absent",
          "counts",
          "min_counts");
  private static final Set<String> MODEL_MEMBERS =
      Set.of("fixture", "fixtures", "language", STRICT_CONFORMANCE);

  private Scenarios() {}

  /**
   * Every scenario in a directory of scenario files.
   *
   * @param directory the {@code scenarios} directory
   * @return the scenarios, in file then declaration order
   */
  static List<Scenario> load(Path directory) {
    List<Path> files;
    try (Stream<Path> entries = Files.list(directory)) {
      files =
          entries
              .filter(path -> path.getFileName().toString().endsWith(".json"))
              .sorted(Comparator.comparing(Path::toString))
              .toList();
    } catch (IOException e) {
      throw new UncheckedIOException("reading " + directory, e);
    }
    if (files.isEmpty()) {
      throw new IllegalArgumentException("no scenario files in " + directory);
    }

    List<Scenario> scenarios = new ArrayList<>();
    Map<String, String> seen = new HashMap<>();
    for (Path file : files) {
      JsonObject suite = read(file).getAsJsonObject();
      for (JsonElement element : suite.getAsJsonArray("scenarios")) {
        Scenario scenario = scenario(element.getAsJsonObject(), file);
        String duplicate = seen.put(scenario.id(), file.toString());
        if (duplicate != null) {
          throw new IllegalArgumentException(
              file + ": scenario id " + scenario.id() + " is already declared in " + duplicate);
        }
        scenarios.add(scenario);
      }
    }
    return List.copyOf(scenarios);
  }

  private static JsonElement read(Path file) {
    try {
      return JsonParser.parseString(Files.readString(file, StandardCharsets.UTF_8));
    } catch (IOException e) {
      throw new UncheckedIOException("reading " + file, e);
    }
  }

  private static Scenario scenario(JsonObject object, Path file) {
    reject(object, SCENARIO_MEMBERS, file + ": scenario");
    String id = text(object, "id");
    String rpc = text(object, "rpc");
    if (id.isEmpty() || rpc.isEmpty()) {
      throw new IllegalArgumentException(file + ": every scenario needs an id and an rpc");
    }
    return new Scenario(
        id,
        text(object, "description"),
        rpc,
        strings(object.get("requires_capabilities")),
        object.has(EXPECT_WITHOUT_CAPABILITY)
            ? Optional.of(expect(object.getAsJsonObject(EXPECT_WITHOUT_CAPABILITY), file))
            : Optional.empty(),
        object.has(MODEL)
            ? Optional.of(fixture(object.getAsJsonObject(MODEL), file))
            : Optional.empty(),
        object.has(REQUEST) ? object.getAsJsonObject(REQUEST) : new JsonObject(),
        object.has(EXPECT) ? expect(object.getAsJsonObject(EXPECT), file) : empty(),
        file.toString());
  }

  private static Scenario.Fixture fixture(JsonObject model, Path file) {
    reject(model, MODEL_MEMBERS, file + ": model");
    return new Scenario.Fixture(
        text(model, "fixture"),
        strings(model.get("fixtures")),
        text(model, "language"),
        model.has(STRICT_CONFORMANCE) && model.get(STRICT_CONFORMANCE).getAsBoolean());
  }

  private static Scenario.Expect expect(JsonObject object, Path file) {
    reject(object, EXPECT_MEMBERS, file + ": expect");
    return new Scenario.Expect(
        text(object, "status"),
        text(object, "status_message_contains"),
        object.has(RESPONSE) ? Optional.of(object.getAsJsonObject(RESPONSE)) : Optional.empty(),
        textMap(object.get("contains")),
        stringsMap(object.get("contains_all")),
        strings(object.get("non_empty")),
        strings(object.get("absent")),
        numberMap(object.get("counts")),
        numberMap(object.get("min_counts")));
  }

  private static Scenario.Expect empty() {
    return new Scenario.Expect(
        "", "", Optional.empty(), Map.of(), Map.of(), List.of(), List.of(), Map.of(), Map.of());
  }

  private static void reject(JsonObject object, Set<String> known, String where) {
    for (String member : object.keySet()) {
      if (!known.contains(member)) {
        throw new IllegalArgumentException(where + ": unknown member " + member);
      }
    }
  }

  private static String text(JsonObject object, String member) {
    return object.has(member) ? object.get(member).getAsString() : "";
  }

  private static List<String> strings(JsonElement element) {
    if (element == null) {
      return List.of();
    }
    List<String> values = new ArrayList<>();
    for (JsonElement item : element.getAsJsonArray()) {
      values.add(item.getAsString());
    }
    return List.copyOf(values);
  }

  private static Map<String, String> textMap(JsonElement element) {
    if (element == null) {
      return Map.of();
    }
    Map<String, String> values = new LinkedHashMap<>();
    for (Map.Entry<String, JsonElement> entry : element.getAsJsonObject().entrySet()) {
      values.put(entry.getKey(), entry.getValue().getAsString());
    }
    return Map.copyOf(values);
  }

  private static Map<String, List<String>> stringsMap(JsonElement element) {
    if (element == null) {
      return Map.of();
    }
    Map<String, List<String>> values = new LinkedHashMap<>();
    for (Map.Entry<String, JsonElement> entry : element.getAsJsonObject().entrySet()) {
      JsonArray items = entry.getValue().getAsJsonArray();
      values.put(entry.getKey(), strings(items));
    }
    return Map.copyOf(values);
  }

  private static Map<String, JsonElement> numberMap(JsonElement element) {
    if (element == null) {
      return Map.of();
    }
    return Map.copyOf(element.getAsJsonObject().asMap());
  }
}
